### ISSUE TITLE
Doc Switcher Inconsistent Background Repaint

### ISSUE THREAD

**Issue by Christopher-Hayes** [notepad-plus-plus/notepad-plus-plus#2937]
### Description of the Issue
When Doc Switcher is set in vertical mode on the left side it inherits the background color of the editor. But, in certain cases, it doesn't appear to repaint when it should. This gives the doc switcher the same light grey background as the title bar.  Happens in Windows 10 when the user switches to task view.

### Steps to Reproduce the Issue
1. Open Notepad++
2. Switch to Task View ( Windows Button + Tab )
3. Exit Task View

### Expected Behavior
The doc switcher should repaint the background upon refocus.

### Actual Behavior
Appears the doc switcher does not repaint, producing the light grey background of the title bar.

### Debug Information
Notepad++ v7.2.2   (32-bit)
Build time : Nov 27 2016 - 16:31:07
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 10
Plugins : AutoSave.dll mimeTools.dll NppConverter.dll NppExport.dll NppFTP.dll PluginManager.dll

**Comment by SinghRajenM** [comment]
I confirm the behavior. It is applicable for all the dockable items such as doc switcher, function list, folder as workspace. etc.

Notepad++ v7.3.2   (32-bit)
Build time : Feb 12 2017 - 23:15:39
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 10 (64-bit)
Plugins : DSpellCheck2.dll mimeTools.dll NppConverter.dll NppExport.dll PluginManager.dll 


![image](https://cloud.githubusercontent.com/assets/14791461/23179871/6c1300e6-f895-11e6-85c0-57ef02193eda.png)

**Comment by jrmain** [comment]
Confirmed here as well. It happens for the Project pane, the Functions list, and the Favorites pane. The easiest way to reproduce it that I've found is to run NP++, make sure that one of the problematic panes is displayed, then press Windows-E to display Windows Explorer. Even if the Explorer window appears on a different display, the background of the NP++ pane changes colour.

Notepad++ v7.4.1 (32-bit)
OS : Windows 8.1 (64-bit)

**Comment by pborunda** [comment]
I see the same issue. It's very distracting. Especially when I set the default background to black.
Notepad++ v7.5.1 (32-bit)

**Comment by CurbedLarry** [comment]
Still an issue in 7.7.1 and also happens after unlocking screen (Notepad++ was in focus when screen locked).

Quick way to force a repaint is minimise and maximise by clicking taskbar icon/bar twice.

Don't remember seeing it on Windows 7 at all, so appears to be an issue with 8 and 10.

**Comment by Tapon94** [comment]
Idem in 7.7.1 with theme vs2015-dark (on windows 10).
Any solution ?

**Comment by ArkadiuszMichalski** [comment]
I can confirm this also in Win7 (is applicable for all the dockable panels, like we see here https://github.com/notepad-plus-plus/notepad-plus-plus/issues/2937#issuecomment-281441991-permalink), but it occurs randomly and I haven't figured out the reproduction steps yet.

**Comment by pryrt** [comment]
another instance: https://community.notepad-plus-plus.org/topic/21751/folder-as-workspace-in-dark-mode-has-the-problem-when-open-file-explorer

and https://community.notepad-plus-plus.org/topic/21350/folder-as-workspace-problems may have been the same, too.

**Comment by thomassmith311** [comment]
In addition, right-clicking a file in the file explorer and hit "Open with Notepad++" opens Notepad++ with the colors already inverted.

**Comment by ArkadiuszMichalski** [comment]
@donho  I can't reproduce this error only with Windows (for me Explorer works fine). Maybe it happend after some sleep, but it's hard to repeat. Today I accidentally discovered a program that always reproduces this problem. STR:

1. Download [portable uTorrent.](https://portableapps.com/apps/internet/utorrent_portable) and extract.
2. Run Notepad++ and open some panel, like `Function list` or `Folder as Workspace`.
3. Run uTorrent.
4. Switch to Notepad++. It should have this background problem in panel. If not then switch to uTorrent and press minimize window button (in top-right corner) and bring this window back to the front, then switch back to Notepad++.
5. Background in Notepad++ can be fixed by minimize it window and bring it back to the front. 
6. We can repeat step 4 and 5 how much we want.

It seems that other programs are breaking Notepad ++ again, just like [Slack in other bug](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9717)?

Btw, now I see that `Document List` panel doesn't have this problem, as opposed to, for example, `Function list` or `Folder as Workspace`.

![image](https://user-images.githubusercontent.com/2730894/143921088-1e635cd4-44b7-43f5-8d0b-657a5344351f.png)

**Comment by donho** [comment]
@ArkadiuszMichalski 
Can I download here instead?
https://www.utorrent.com/web/compare/
The basic one with AD, The AD is in utorrent but not on the browser, could you confirm it?

**Comment by ArkadiuszMichalski** [comment]
Iif you prefer to install it, you can download it directly from the website, but choose the [Classic version](https://www.utorrent.com/desktop/),

**Comment by donho** [comment]
@ArkadiuszMichalski 
Thank you for the info. I found out a way to reproduce this issue steadily without utorrent:
I put Notepad++ ( in dark mode +FunctionList) on the background by putting another app on the foreground, then I put ma laptop in sleep.
After waking up PC, I can see the snapshot of Notepad++ with the white bg FunctionList and click it to show as snapshot.

**Comment by donho** [comment]
It's rather a bug of Windows' API than the bug of Notepad++. The visual glitch comes from TreeView (FunctionList, Folder As WorkSpace, Project Panels etc ...) but ListView (ASCII Insertion Panel & Document List) works fine.

I have done some fixes about this issue. However, it's not perfect when use mouse to do the snapshot:

![image](https://user-images.githubusercontent.com/90293/144102568-b7229942-9f9d-4e3e-bb77-3c0011b74fc9.png)

OTOH, once the click done to show Notepad++, it becomes normal:

![image](https://user-images.githubusercontent.com/90293/144104442-50fc35e0-145a-48c6-95f4-b48dad7ebec6.png)

That's the best fix I can do for this bug.

**Comment by shriprem** [comment]
@donho , a comprehensive fix for this bug is quite possible. The following code that handles [WM_SYNCPAINT](https://docs.microsoft.com/en-us/windows/win32/gdi/wm-syncpaint#remarks) needs to be added inside `Notepad_plus::process` function within [NppBigSwitch.cpp](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/NppBigSwitch.cpp).

```C++
LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam) {
...
		case WM_SYNCPAINT:
			RedrawWindow(hwnd, NULL, NULL, RDW_INVALIDATE | RDW_ERASE | RDW_ALLCHILDREN);
			break;
...
}
```

In addition, the 45 or so lines of current code under `case WM_ACTIVATE` can be safely removed, and the bug will still remain fixed.

**Reasoning behind this proposed code**:
The main NPP top-level window has the `WS_CLIPCHILDREN` style. The various docking panes are children of this top-level window. Because of the `WS_CLIPCHILDREN` style at the top-level window, by default, the refresh paint call is not percolating down to the child windows (i.e., the docking panes). So, the above code handling the [WM_SYNCPAINT](https://docs.microsoft.com/en-us/windows/win32/gdi/wm-syncpaint#remarks) message is going to ensure that a redraw is invoked on the child windows. `RDW_ALLCHILDREN` is the critical flag in the [RedrawWindow](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-redrawwindow) call.

I have successfully tested this fix with both 32-bit and 64-bit builds of NPP on Windows 10 and Windows 7 PCs.
* With this fix, no background clipping whatsoever will be observed on the _Folder as Workspace_ panel at any point of time. Thus, there is no need to click once on the NPP window (as suggested in the preceding post) to invoke the `WM_ACTIVATE` code behind the scenes to redraw the panel, and thereby clear its background clipping.
* With this fix, when new panels are added to NPP in the future, there will be no need for additional code to handle their redraws under `WM_ACTIVATE`.
* With this fix, darkmode refresh will be handled for plugin panels also. I have tested this to be working in darkmode with my own NPP plugins: [FWDataViz](https://github.com/shriprem/FWDataViz) and [GoToLineCol](https://github.com/shriprem/Goto-Line-Col-NPP-Plugin).

**A more convenient way to trigger background clipping in older NPP versions, and test the fix with the proposed code**
Just download the tiny [Process Explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) utility from the Microsoft SysInternals site. While an NPP version prior to 8.1.9.3 is running with various side panels, launch Process Explorer either via _Windows Start -> Run_ or the NPP's own _Run_ menu option. I think this is being possible with Process Explorer because the opacity of its window can be set anywhere between _Opaque_ to 10% (look under its menu _View_ » _Opacity_).

The screenshot below was captured with the NPP v8.1.9.3 release. In the screenshot:
* The _Folder as Workspace_ panel on the left and the _Goto Line Column_ panel at the bottom have their backgrounds clipped due to the launch of Process Explorer in the foreground. The background clipping is happening because both these panels are children of the top-level NPP window.
* However, the _Preferences_ popup dialog of the _Goto Line Column_ plugin does not have its background clipped because it is an independent top-level window that is receiving its own `WM_SYNCPAINT` message following the launch of Process Explorer. It was this observation, gleaned via the VS Spy++ utility, that lead me to the proposed code fix.
[_Note: The clipping artifacts seen in the screenshot below will not occur with the proposed code fix._]
![image](https://user-images.githubusercontent.com/45252729/145748014-a6b1f3b5-afcd-4788-b914-58fd0b6a809e.png)

**Comment by xomx** [comment]
The fix works for me (8.1.9.3 x64, Win8.1Ent).
